programs/stake/src/stake_state.rs

```diff
@@ -1030,49 +1042,50 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
 
     fn merge(
         &self,
-        source_stake: &KeyedAccount,
+        source_account: &KeyedAccount,
         clock: &Clock,
         stake_history: &StakeHistory,
         signers: &HashSet<Pubkey>,
     ) -> Result<(), InstructionError> {
-        if source_stake.owner()? != id() {
+        if source_account.owner()? != id() {
             return Err(InstructionError::IncorrectProgramId);
         }
 
-        let meta = match self.state()? {
-            StakeState::Stake(meta, stake) => {
-                // stake must be fully de-activated
-                if stake.stake(clock.epoch, Some(stake_history), true) != 0 {
-                    return Err(StakeError::MergeActivatedStake.into());
-                }
-                meta
-            }
-            StakeState::Initialized(meta) => meta,
-            _ => return Err(InstructionError::InvalidAccountData),
-        };
+        let (meta, stake) = get_info_if_mergeable(self, clock, stake_history)?;
+
         // Authorized staker is allowed to split/merge accounts
         meta.authorized.check(signers, StakeAuthorize::Staker)?;
 
-        let source_meta = match source_stake.state()? {
-            StakeState::Stake(meta, stake) => {
-                // stake must be fully de-activated
-                if stake.stake(clock.epoch, Some(stake_history), true) != 0 {
-                    return Err(StakeError::MergeActivatedStake.into());
-                }
-                meta
-            }
-            StakeState::Initialized(meta) => meta,
-            _ => return Err(InstructionError::InvalidAccountData),
-        };
+        let (source_meta, source_stake) =
+            get_info_if_mergeable(source_account, clock, stake_history)?;
 
         // Meta must match for both accounts
         if meta != source_meta {
```

Comment:
Considering,the current state of affair of two variants of rent_exempt_reserve even after stake rewrite, should we relax `rent_exempt_reserve` as well?

Reply:
Sure!  I don't see that causing any harm.

This won't complicate the `Delegation::stake` value fix rewrites, since support for both will be enabled at the same time, right?